### PR TITLE
feat: Feature tag component icon prop support

### PR DESCRIPTION
**Pull request by vtsybulin**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] TypeScript definition update
- [ ] Test Case
- [ ] Branch merge
- [x] Other (New component prop)

### 💡 Background and solution

Hi. This one aims to update the Tag component to support the 'icon' prop in alignment with what the Button component currently offers.

Final demo:
72070175-8a7a4f00-32f1-11ea-82e7-eb1382610f9b

Usage example:

`import { Tag } from 'antd';`
`import { TwitterOutlined } from '@ant-design/icons';`
`...`
`<Tag icon={<TwitterOutlined />}>Twitter Tag</Tag>`

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

-----

-----
[View rendered components/tag/demo/icon.md](https://github.com/vtsybulin/ant-design/blob/feature-tag-component-icon-prop-support/components/tag/demo/icon.md)
[View rendered components/tag/index.en-US.md](https://github.com/vtsybulin/ant-design/blob/feature-tag-component-icon-prop-support/components/tag/index.en-US.md)
[View rendered components/tag/index.zh-CN.md](https://github.com/vtsybulin/ant-design/blob/feature-tag-component-icon-prop-support/components/tag/index.zh-CN.md)

### CODE REVIEW

File: components/tag/__tests__/index.test.js [hunk 1]
```diff
@@ -52,6 +52,17 @@ describe('Tag', () => {
...
+  it('should warning when pass a string as icon props', () => {
+    const warnSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
+    mount(<Tag icon="ab" />);
+    expect(warnSpy).not.toHaveBeenCalled();
+    mount(<Tag icon="twitter" />);
+    expect(warnSpy).toHaveBeenCalledWith(
+      `Warning: [antd: Tag] \`icon\` is using ReactNode instead of string naming in v4. Please check \`twitter\` at https://ant.design/components/icon`,
+    );
+    warnSpy.mockRestore();
+  });
```
Comment (afc163) [inline components/tag/__tests__/index.test.js]: Don't need this now.

File: components/tag/style/index.less [hunk 2]
```diff
@@ -119,6 +119,12 @@
   .make-status-color-classes('cyan', processing);
   .make-status-color-classes('red', error);
   .make-status-color-classes('orange', warning);
+
+  // To ensure that a space will be placed between character and `Icon`.
+  > .@{iconfont-css-prefix} + span,
+  > span + .@{iconfont-css-prefix} {
+    margin-left: 8px;
```
Comment (afc163) [inline components/tag/style/index.less]:
```suggestion
    margin-left: 7px;
```

same as Tag padding.

**Comment by afc163** [comment]
rebase master

**Comment by vtsybulin** [comment]
@afc163 Addressed requested changes. There are no changes to be applied by rebasing, as I can see tho

File: components/tag/index.tsx [hunk 3]
```diff
@@ -106,24 +108,39 @@ class Tag extends React.Component<TagProps, TagState> {
...
+    );
+    const iconNode = icon || null;
+    const kids = iconNode ? (
+      <>
+        {iconNode}
+        <span>{children}</span>
+      </>
+    ) : (
+      children
+    );
```
Comment (afc163) [inline components/tag/index.tsx]: So is this.

File: components/tag/index.zh-CN.md [hunk 4]
```diff
@@ -23,6 +23,7 @@ title: Tag
 | color | 标签色 | string | - |
 | onClose | 关闭时的回调 | (e) => void | - |
 | visible | 是否显示标签 | boolean | `true` |
+| icon | 设置图标类型 | ReactNode | - |  |
```
Comment (afc163) [inline components/tag/index.zh-CN.md]:
```suggestion
| icon | 设置图标 | ReactNode | - |  |
```

File: components/tag/index.tsx (hunk 3 above)
Comment (vtsybulin) [reply components/tag/index.tsx]: Sorry, not sure what u mean

Comment (vtsybulin) [reply components/tag/index.tsx]: @afc163 🔝

Comment (afc163) [reply components/tag/index.tsx]: We don't need this either, cause antd v3 never supported Tag[icon].

File: components/tag/index.tsx [hunk 5]
```diff
@@ -106,24 +108,39 @@ class Tag extends React.Component<TagProps, TagState> {
...
-    const { children, ...otherProps } = this.props;
+    const { children, icon, ...otherProps } = this.props;
     const isNeedWave =
       'onClick' in otherProps || (children && (children as React.ReactElement<any>).type === 'a');
     const tagProps = omit(otherProps, ['onClose', 'color', 'visible', 'closable', 'prefixCls']);
+    warning(
+      !(typeof icon === 'string' && icon.length > 2),
+      'Tag',
+      `\`icon\` is using ReactNode instead of string naming in v4. Please check \`${icon}\` at https://ant.design/components/icon`,
+    );
```
Comment (afc163) [inline components/tag/index.tsx]: Sorry for reply in wrong line.

File: components/tag/index.tsx (hunk 3 above)
Comment (vtsybulin) [reply components/tag/index.tsx]: Oh, that what you meant :) Fixed. Thought you somehow related to the rebase thing

**Comment by afc163** [comment]
Oh, feature branch please~

**Comment by vtsybulin** [comment]
@afc163 done

**Comment by afc163** [comment]
Rebase feature

**Comment by vtsybulin** [comment]
@afc163 done

**Comment by afc163** [comment]
Oops, mixed other commits don't blong to your PR.

**Comment by vtsybulin** [comment]
I was planning just to `git rebase feature` 🤔 😕

**Comment by vtsybulin** [comment]
@afc163 can you help with that please, so I don't fuck up again 🤔😄

**Review by afc163 — APPROVED** [review]

**Comment by vtsybulin** [comment]
@afc163 cool, thanks 💪

**Pull request merged**